src/libraries/Microsoft.PowerFx.Core/Public/Engine.cs

```diff
@@ -163,14 +163,21 @@ private protected INameResolver CreateResolverInternal(out ReadOnlySymbolTable s
             symbols = ReadOnlySymbolTable.Compose(localSymbols, GetCombinedEngineSymbols());
             return symbols;
         }
+               
+        private readonly ComposedSymbolTableCache _functionListCache = new ComposedSymbolTableCache();
 
         /// <summary>
         /// Get a combined engine symbol table, including builtins and config. 
         /// </summary>
         /// <returns></returns>
         public ReadOnlySymbolTable GetCombinedEngineSymbols()
         {
-            var symbols = ReadOnlySymbolTable.Compose(EngineSymbols, SupportedFunctions, Config.SymbolTable, PrimitiveTypes);
+            // Compose all the symbol tables most likely to have functions into a single 
+            // symbol table and then cache that. 
+            // That will cache unifying into a single TexlFunctionSet - which is the most expensive part. 
+            var functionList = _functionListCache.GetComposedCached(SupportedFunctions, Config.SymbolTable);
```

Comment:
This is critical.  This combines the symbol tables that have non-empty function sets into a single table, which can then be cached.  And being a single table, then the Compose() operator above becomes trivial and fast.